text/0000-type-inference.md

```diff
@@ -0,0 +1,168 @@
+    Service: Service
+  ) => Middleware,
+  cleanup?: (service: Service) => Promise<void>,
+|};
+
+function createPlugin<Deps, Service>(p: FusionPlugin<Deps, Service>): FusionPlugin<Deps, Service> {
+  // ...
+}
+```
+
+In this example, the `Service` argument is in the input position in the `cleanup` function as well as the `middleware` function. Despite `deps` being used
+in the `provides` and `middleware` function, it is not directly reachable from exports because of the usage of `$ObjMap`. 
+
+Looking at the `cleanup` method is the simplext example. The reason the `Service` generic here qualifies as a type in the input position is demonstrated by
+the following example:
+
+```js
+export const plugin = createPlugin({
+  provides: () => {
+    return 5;
+  }
+  cleanup(service) => {
+    console.log(service);
+  }
+});
+
+// other file
+import {plugin} from 'file';
+plugin.cleanup('test');
+```
+
+In this case, the `service` argument to `cleanup` is inferred as `number | string` while with fusion semantics it should be `number` and result in an error
+when called with a string. This does not match the flow semantics however due to the process of subtyping ([read more](https://flow.org/en/docs/lang/subtypes/)).
+
+Taking a step back, if we look at our `createPlugin` API we can take advantage of the fact that the return value from `createPlugin` is designed to be opaque.
+In other words, it is not designed to have its properties accessed directly by consumers like in the cleanup example. 
+
+There are two parts of the type checking of fusion plugins. The first part is in the `createPlugin` call itself. This makes sure all the parts of the plugin are 
+consistent. For example, the declared deps match the types of the deps passed into `provides` and `middleware`, and the type returned from `provides` matches the 
+type injected into `cleanup` and `middleware`. All of this can be done without having any return type from `createPlugin`.
+
+The second part is type checking the `app.register` and `app.enhance` calls. This is done to verify the type of the token matches the provides type of the plugin, 
+and this is the part which depends on the return type of `createPlugin`. The key thing to notice here is that we only need the type of the `provides` method of 
+the plugin to verify it matches the token. At this point, deps, cleanup, and middleware are all irrelevant. 
+
+Because `createPlugin` should return an opaque type and because we only need the `provides` type to verify with `app.register`, we can update the `createPlugin` 
+return type to remove all generics in input positions without losing any type safety for consumers. There are several ways we could do this, but the simplest
+would be replacing the `service` generic with `any` in the return type of `createPlugin` whenever it is referenced in an input position. We also need to specify
+the properties as covariant.
+
+```js
+type CreatePluginArgs<+Deps, +Service> = {|
```

Comment:
Does this work?  I think the generic declaration for `Service` here might need to be invariant as it is still being passed in as an input.

Reply:
Yea you are correct - copy paste error